Merge Galaxy-XSD into project

jmchilton opened the pull request:
Transplanted with commit authors maintained.

The most obvious advantages of it living here instead of its own project is that it gives it higher profile and it allows the version of the schema to match that of Galaxy (so we can check at least if the XML structure of a tool is valid for a given release). 

This has a few more advantages being in core that will probably come from follow up commits if this makes it in:
- Allow docs for tool XML generated from this to live in this project.
- Allow distribution with galaxy-lib.
- Allow checking testing tools as part of CI, ensures all new tool features are added to XSD and documented.

A Markdown document generated from this XSD is available for preview at https://github.com/jmchilton/planemo/blob/xsd_and_docs/docs/schema.md#tool%7Ctests%7Ctest%7Crepeat. I think in a subsequent commit I'll move those docs here - but for now it is a good way to "visualize" the XSD.

jmchilton committed:
Deleting directories from SVN that should have been ignored.

git-svn-id: https://svn.msi.umn.edu/tint/trunk@427 29e1c259-b37a-0410-ae23-5fb2df606144

jmchilton committed:
Work on galaxy support, implemented repeat blocks in the UI.

JeanFred committed:
Add basic ReadMe

JeanFred committed:
Remove custom namespaces

JeanFred committed:
Change tool element to <xs:all>

Element order is not enforced by Galaxy.

JeanFred committed:
Add <requirements> tag set

This includes:
-<requirements>
-<requirement>
-Enumeration RequirementType

JeanFred committed:
Complete <repeat> tag set

Add attributes min and max to Repeat element

JeanFred committed:
Add <options> tag to <param>

The <param> tag can contain <options> to dynamically create a list of options.

JeanFred committed:
Add <label> tag to <param>

JeanFred committed:
Add missing attributes to <param> tag

Add seven missing attributes to <param> tag
Add the necessary datatypes for display, hierarchy and checked.

JeanFred committed:
Add missing attributes and element to <data> tag

Add three missing attributes to <data> tag
Add <filter> element to sequence

JeanFred committed:
Fix indentation

JeanFred committed:
Add <parallelism> tag

Add the <parallelism> tag (which serves for splitting job)
and the datatype for the 'method' attribute.

JeanFred committed:
Add <stdio> tag set to <tool>

Add <stdio> tag set, to handle tool messages on standard output
and standard error.

- Add <stdio> tag
- Add <regex> and <exit_code> sub elements
- Add datatypes for 'range', 'level' and 'source' attributes

JeanFred committed:
Change type for requirement version

Version information can be too complex for xs:decimal
(eg "0.12.7"). Change for xs:string.

JeanFred committed:
Add column, filter and validator tags

JeanFred committed:
Add ftype attribute to param tag in tests

JeanFred committed:
Add <code> tag set

JeanFred committed:
Remove required use from <requirement> version

JeanFred committed:
Edit <options> tag set

JeanFred committed:
Add <validator> to <param>

JeanFred committed:
Add 'split' attribute to <validator>

JeanFred committed:
Add <actions> tag set

JeanFred committed:
Extend CheckedType to accept unspecified values

According to the documentation, CheckedType can either
be 'yes', 'true' or 'on' ; but many tools use '[Ff]alse'
which seems to be accepted/ignored by Galaxy.

JeanFred committed:
Add PermissiveBoolean datatype

<xs:boolean> can either be 'true', 'false', '1' or 'O' ; but many
Galaxy tools use 'True' and 'False' (capitalized), which appears
to be ignored/accepted by Galaxy.
This PermissiveBoolean type is meant to be used in place of xs:boolean
when needed.

JeanFred committed:
Add attributes 'ftype' and 'sort' to test <output>

These are undocumented attributes.

JeanFred committed:
Add (undocumented) attributes to <filter>

Add attributes 'compare', 'keep' and 'value' to <filter> tag
(though these attributes are not documented).
Add CompareType for possible 'compare' values.

JeanFred committed:
Edit filter tag for output tag

JeanFred committed:
Rewrite <actions> handling

JeanFred committed:
Add (undocumented) 'dbkey' attribute to TestParam

JeanFred committed:
Add 'from_file' value to ActionsOptionType

JeanFred committed:
Use PermissiveBoolean for ParamOption 'selected'

JeanFred committed:
Add attributes tool_type and URL_method to <tool>

And corresponding enumeration Datatypes.

JeanFred committed:
use PermissiveBoolean on two attributes of <param>

JeanFred committed:
Add (undocumented) value 'no' to CheckedType

JeanFred committed:
Add (undocumented) <uihints> to <tool>

See /lib/galaxy/tools/__init__.py:1069 for usage

JeanFred committed:
Add (undocumented) attributes to <inputs>

Add attributes 'action', 'check_values', 'method' and 'target'
to <inputs>.
Add relevant Target datatype.

JeanFred committed:
Add <sanitizer> tag

JeanFred committed:
Add (undocumented) attribute metadata_name to <param>

JeanFred committed:
Add (undocumented) attributes to <options>

Add 'options_filter_attribute' attribute
(see lib/galaxy/tools/parameters/basic.py)
Add 'transform_lines' attribute
(see lib/galaxy/tools/parameters/dynamic_options.py)
transform_lines and

JeanFred committed:
Add (undocumented) 'ref_attribute' to <filter>

Add attribute 'ref_attribute' to <filter>
(see lib/galaxy/tools/parameters/dynamic_options.py:153)

JeanFred committed:
Add (undocumented) 'startswith' to <options>

Add attribute 'startswith' to <options> tag
(see lib/galaxy/tools/parameters/dynamic_options.py:416)

JeanFred committed:
Add (undocumented) attribute 'display' to <inputs>

See lib/galaxy/tools/__init__.py:1531

JeanFred committed:
Fix value baseurl in <param> 'type' attribute

JeanFred committed:
Fix type of min and max in <validator>

lindenb committed:
added doc and JXB bindings

lindenb committed:
updated README

JeanFred committed:
Add Continuous integration with Travis for XSD linting

At the very least these XSD files should be XML valid.

JeanFred committed:
Fix dependency installation for Travis

Build was failing with
$ apt-get install libxml2-utils
E: Could not open lock file /var/lib/dpkg/lock - open (13: Permission denied)
E: Unable to lock the administration directory (/var/lib/dpkg/), are you root?

APT dependencies must be installed as sudoer.

JeanFred committed:
Add Travis badge to ReadMe

jmchilton committed:
Add me (@jmchilton) to authors list.

My name is on the earliest commits (thanks for preserving that).

jmchilton committed:
Fix CitationType attribute.

Add Galaxy's test tool for citations to test.urls.

jmchilton committed:
Update XSD for container tags.

Add Galaxy test tool for Docker container tags to test.urls.

jmchilton committed:
Update XSD for very basic data collection usage.

 - Allow 'data_collection' parameters and allow them to specify a 'collection_type'.
 - Allow test parameters to be flat collections.
 - Add test/functional/tools/collection_paired_test.xml to test.urls to check all of this.

Lots more TODO for recent tool collection support - output collections definitions, output collection tests, dynamic discovery tags, specifying nested collections for test inputs.

jmchilton committed:
Drop embeddedValue for test stuff.

These were an extension to the Galaxy XML syntax for TINT - which is no longer actively maintained.

jmchilton committed:
Start work on implementing assert_contents tag in tests.

Just enough to get Galaxy's simple_contructrs.xml to validate - add this to test.urls. More work should be done on these - but it should be automated derived from Galaxy somehow.

JeanFred committed:
Add Shell test script

Following the idea of @lindenb in the Makefile.

This tests all the given URLs in test.urls and
outputs a compliant TAP.

JeanFred committed:
Add MIT license

Per agreement of all authors.

Closes #2.

JeanFred committed:
Expand history in ReadMe

JeanFred committed:
Fix authors in License file

Follow up on #2.

mr-c committed:
first try

jmchilton committed:
Improvements to XSD:

- Implement section.
- Implement and document request_parameter_translation, request_parameter and its children (value_translation and append_param).
- Implement detect_errors on the command block and document defaults.

jmchilton committed:
More progress...

- Fix to previous commit.
- Remove downstream config file type that was abandoned work and not in galaxy-core.
- More documentation for configfile stuff.
- Improvements to representation for output and santizer.

jmchilton committed:
Simplify schema and improve docs related to sanitization.

jmchilton committed:
Improve docs on filter and change_format of outputs.

jmchilton committed:
And more progress...

- Docs for help.
- Docs for tests.
- Docs for code files.
- Implementing more of the assert_xxx test flags.
- Improved coverage of test assertion element descriptions.

jmchilton committed:
More progress...

- Refine test outputs and include syntax for testing extra files and md5sums.

jmchilton committed:
More progress.

- Add strict to command.
- Add conditional and repeat to test params.
- Add count to output collections.
- Implement more collection output attributes (type_source, etc...).
- Reduce need for type on collection defs (since type_source can be defined).

jmchilton committed:
More progress...

- Allow section in test parameters.
- Implement basic metadata testing.

jmchilton committed:
More progress...

- Allow nested elements in test output collections (hackily)
- Implement ``compare`` and ``delta`` for test output comparison.
- Implement the new ``checksum`` output test.

jmchilton committed:
More progress...

- Add tool action.
- Stub workflow_compatible and profile_version attributes on tool.
- Add many more attributes to parallelism tag.
- Add maxseconds and expect_num_outputs to tool test tag.
- Small fix to <repeat> block in test cases.
- Implement assigning metadata in test params.
- Allow more versitle ordering of test blocks (too permissive now).
- Implement testing of output metadata and discovered datasets.
- Add nginx_upload attribute used by upload.xml and various conditional attributes for pulling in values from Galaxy internals.
- Implement the stub of a ``upload_dataset`` param and say not to use it.
- Add color and library data parameter types.
- More attributes to filters and validators.
- Implement auto_format on outputs.
- Implement filters and explicit data outputs on output collections.
- Implement inputs as json element.
- Add empty dataset validator.

jmchilton committed:
Stub out environment variables and add a new tool type.

jmchilton committed:
Embed citations right in galaxy.xsd.

Not that I mind splitting it up - but it will make a couple things easier downstream and those few tags weren't enough to warrent breaking things up like that.

jmchilton committed:
Many more fixes derived from testing tools-iuc.

jmchilton committed:
And more progress - now covers all Galaxy test tools and all IUC.

- Add <options> block.
- Add <trackster_conf> block.
- Add a newerish validator and allow filter/validator columns to be column names.
- More permissives boolean values to match Galaxy in various places.
- Other mostly obsecure features.

jmchilton committed:
Doc updates

jmchilton committed:
Doc improvements.

jmchilton committed:
Much richer documentation for a few tags including ``command``.

jmchilton committed:
More work on documentation - revise last commit and add repeats and sections.

jmchilton committed:
Work on docs.

jmchilton committed:
More docs...

- Added docs for param's option.
- Started work on validators and options, but are WIP though.

jmchilton committed:
More docs for validators.

jmchilton committed:
A lot more documentation around parameter options.

Also fixed a bit of a problem - turns out <filter> in action conditionals is very different than filter in options so I've split these types out.

jmchilton committed:
More documentation...

Cleaned up the stdio tags and help section, flushed out citations and parameter options. Small touchups other places also.

jmchilton committed:
Mention code files as a way to generate dynamic options.

Transplant example from Galaxy wiki of doing this.

jmchilton committed:
Flush out docs for requirements and child tags.

jmchilton committed:
Flush out docs for sanitizers, configfiles, etc...

jmchilton committed:
Working my way through basic parameter documentation.

jmchilton committed:
More work on parameters and environment variables.

jmchilton committed:
Finishing first pass at filling out wiki docs for ``param``.

And other small fixes.

jmchilton committed:
Starting to polish and flush out data outputs documentation.

jmchilton committed:
More docs - conditional output actions and some more on discover datasets.

jmchilton committed:
A bunch of work on cleaning up and flushing out documentation for test tags.

jmchilton committed:
Document the test ``output_collection`` tag.

jmchilton committed:
Fix missing annotation.

jmchilton committed:
Revise automated testing, check against test tools and IUC tools.

jmchilton committed:
Fixes of structure broken with docs.

jmchilton committed:
Use the new action conditional filter.

jmchilton committed:
Add a fictious XML schema with Galaxy stuff.

New attributes:

 - deprecated
 - best_practices (provides links of the form http://planemo.readthedocs.io/en/latest/standards/docs/best_practices/tool_xml.html#XXX)

jmchilton committed:
Remove embedded best practice stuff - the help will have links now.

jmchilton committed:
Fill out some docs for test sections...

jmchilton committed:
Document testing discovered datasets and metadata.

jmchilton committed:
Fix a syntax error.

jmchilton committed:
Update the link for multiple outputs.

jmchilton committed:
Relocate Galaxy XSD files into Galaxy directory structure.

jmchilton committed:
Cleanup XSD for move to Galaxy core.

jmchilton committed:
Merge branch 'master' of Galaxy-XSD into 'dev'

martenson commented:
+1 on the principle of shipping xsd with Galaxy

On Thu, Sep 15, 2016, 00:23 John Chilton notifications@github.com wrote:

> Transplanted with commit authors maintained.
> 
> The most obvious advantages of it living here instead of its own project
> is that it gives it higher profile and it allows the version of the schema
> to match that of Galaxy (so we can check at least if the XML structure of a
> tool is valid for a given release).
> 
> This has a few more advantages being in core that will probably come from
> follow up commits if this makes it in:
> - Allow docs for tool XML generated from this to live in this project.
> - Allow distribution with galaxy-lib.
> - Allow checking testing tools as part of CI, ensures all new tool
>   features are added to XSD and documented.
> 
> A Markdown document generated from this XSD is available for preview at
> https://github.com/jmchilton/planemo/blob/xsd_and_docs/docs/schema.md#tool%7Ctests%7Ctest%7Crepeat.
> I think in a subsequent commit I'll move those docs here - but for now it
> 
> ## is a good way to "visualize" the XSD.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/galaxyproject/galaxy/pull/2923
> Commit Summary
> - Deleting directories from SVN that should have been ignored.
> - Work on galaxy support, implemented repeat blocks in the UI.
> - Add basic ReadMe
> - Remove custom namespaces
> - Change tool element to xs:all
> - Add <requirements> tag set
> - Complete <repeat> tag set
> - Add <options> tag to <param>
> - Add <label> tag to <param>
> - Add missing attributes to <param> tag
> - Add missing attributes and element to <data> tag
> - Fix indentation
> - Add <parallelism> tag
> - Add <stdio> tag set to <tool>
> - Change type for requirement version
> - Add column, filter and validator tags
> - Add ftype attribute to param tag in tests
> - Add <code> tag set
> - Remove required use from <requirement> version
> - Edit <options> tag set
> - Add <validator> to <param>
> - Add 'split' attribute to <validator>
> - Add <actions> tag set
> - Extend CheckedType to accept unspecified values
> - Add PermissiveBoolean datatype
> - Add attributes 'ftype' and 'sort' to test <output>
> - Add (undocumented) attributes to <filter>
> - Edit filter tag for output tag
> - Rewrite <actions> handling
> - Add (undocumented) 'dbkey' attribute to TestParam
> - Add 'from_file' value to ActionsOptionType
> - Use PermissiveBoolean for ParamOption 'selected'
> - Add attributes tool_type and URL_method to <tool>
> - use PermissiveBoolean on two attributes of <param>
> - Add (undocumented) value 'no' to CheckedType
> - Add (undocumented) <uihints> to <tool>
> - Add (undocumented) attributes to <inputs>
> - Add <sanitizer> tag
> - Add (undocumented) attribute metadata_name to <param>
> - Add (undocumented) attributes to <options>
> - Add (undocumented) 'ref_attribute' to <filter>
> - Add (undocumented) 'startswith' to <options>
> - Add (undocumented) attribute 'display' to <inputs>
> - Fix value baseurl in <param> 'type' attribute
> - Fix type of min and max in <validator>
> - added doc and JXB bindings
> - updated README
> - Add Continuous integration with Travis for XSD linting
> - Fix dependency installation for Travis
> - Add Travis badge to ReadMe
> - Add me (@jmchilton) to authors list.
> - Fix CitationType attribute.
> - Add Shell test script
> - Update XSD for container tags.
> - Update XSD for very basic data collection usage.
> - Drop embeddedValue for test stuff.
> - Start work on implementing assert_contents tag in tests.
> - Add MIT license
> - Expand history in ReadMe
> - Fix authors in License file
> - first try
> - Improvements to XSD:
> - More progress...
> - Simplify schema and improve docs related to sanitization.
> - Improve docs on filter and change_format of outputs.
> - And more progress...
> - More progress...
> - More progress.
> - More progress...
> - More progress...
> - More progress...
> - Stub out environment variables and add a new tool type.
> - Embed citations right in galaxy.xsd.
> - Many more fixes derived from testing tools-iuc.
> - And more progress - now covers all Galaxy test tools and all IUC.
> - Doc updates
> - Doc improvements.
> - Much richer documentation for a few tags including `command`.
> - More work on documentation - revise last commit and add repeats and
>   sections.
> - Work on docs.
> - More docs...
> - More docs for validators.
> - A lot more documentation around parameter options.
> - More documentation...
> - Mention code files as a way to generate dynamic options.
> - Flush out docs for requirements and child tags.
> - Flush out docs for sanitizers, configfiles, etc...
> - Working my way through basic parameter documentation.
> - More work on parameters and environment variables.
> - Finishing first pass at filling out wiki docs for `param`.
> - Starting to polish and flush out data outputs documentation.
> - More docs - conditional output actions and some more on discover
>   datasets.
> - A bunch of work on cleaning up and flushing out documentation for
>   test tags.
> - Document the test `output_collection` tag.
> - Fix missing annotation.
> - Revise automated testing, check against test tools and IUC tools.
> - Fixes of structure broken with docs.
> - Use the new action conditional filter.
> - Add a fictious XML schema with Galaxy stuff.
> - Remove embedded best practice stuff - the help will have links now.
> - Fill out some docs for test sections...
> - Document testing discovered datasets and metadata.
> - Fix a syntax error.
> - Update the link for multiple outputs.
> - Relocate Galaxy XSD files into Galaxy directory structure.
> - Cleanup XSD for move to Galaxy core.
> - Merge branch 'master' of Galaxy-XSD into 'dev'
> 
> File Changes
> - _A_ lib/galaxy/tools/xsd/LICENSE
>   https://github.com/galaxyproject/galaxy/pull/2923/files#diff-0 (22)
> - _A_ lib/galaxy/tools/xsd/README.md
>   https://github.com/galaxyproject/galaxy/pull/2923/files#diff-1 (45)
> - _A_ lib/galaxy/tools/xsd/galaxy.jxb
>   https://github.com/galaxyproject/galaxy/pull/2923/files#diff-2 (22)
> - _A_ lib/galaxy/tools/xsd/galaxy.xsd
>   https://github.com/galaxyproject/galaxy/pull/2923/files#diff-3
>   (4981)
> - _A_ lib/galaxy/tools/xsd/test.sh
>   https://github.com/galaxyproject/galaxy/pull/2923/files#diff-4 (34)
> 
> Patch Links:
> - https://github.com/galaxyproject/galaxy/pull/2923.patch
> - https://github.com/galaxyproject/galaxy/pull/2923.diff
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> https://github.com/galaxyproject/galaxy/pull/2923, or mute the thread
> https://github.com/notifications/unsubscribe-auth/ABuxqmsz9Joq-WOkCZhDcFzt-tSEfvx7ks5qqMhOgaJpZM4J9fDt
> .

bgruening commented:
:+1:

mvdbeek commented:
I think present and future tool authors will be very happy with this. 👍

nsoranzo commented on lib/galaxy/tools/xsd/test.sh:
```diff
@@ -0,0 +1,34 @@
+cd galaxy
+git pull
+cd ..
+
```
I think that lines 2-9 can be removed.

nsoranzo commented on lib/galaxy/tools/xsd/test.sh:
```diff
@@ -0,0 +1,34 @@
+git pull
+cd ..
+
+count=`wc -l tool_files.list | cut -f1 -d' '`
```
There is no `tool_files.list` file.

jmchilton committed:
Address reviewer comments on XSD.

- Merge in @bgruening's numerous awesome fixes.
- Remove test script with issues that make if not a fit for this repo in that form (noted by @nsoranzo).
- Incorporate other changes to the XSD suggested by @nsoranzo.

jmchilton commented:
Pushed babaf7a2c675648e2851f81d985a72321b4eeb44 to address reviewer comments on XSD.
- Merged in @bgruening's numerous awesome fixes from Galaxy-XSD downstream.
- Removed test script with issues that make if not a fit for this repo in that form (noted by @nsoranzo).
- Incorporate other changes to the XSD suggested by @nsoranzo.

bgruening commented on lib/galaxy/tools/xsd/LICENSE:
```diff
@@ -0,0 +1,22 @@
+The MIT License (MIT)
+
+Copyright (c) 2011-2015 John Chilton, Jean-Frédéric Berthelot, Pierre Lindenbaum
```
2016

jmchilton replied:
Updated.

bgruening commented:
Thanks @jmchilton! Ready for merge?

jmchilton committed:
Update the LICENSE for the XSD file.

jmchilton commented:
I consider it ready for merging. I'll open a new PR regarding the LICENSE (let's see if we can drop it as a special license inside of Galaxy) - though I did update the year.

martenson commented:
I would leave it open for a bit - it deserves to be open for more than 14 hours I believe.

jmchilton committed:
Add XSD contributors to CONTRIBUTORS.md.

jmchilton commented:
I'm pretty sure we can relocate this here and keep the LICENSE statement as it is, but it would be cleaner if we re-licensed it under Galaxy's license.

@JeanFred, @lindenb as co-copyright holders I was hoping to get your explicit permission to re-license this under the terms of [Galaxy's license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt). It is a permissive (not copy-left) license very much in the spirit of the original MIT license I believe. Regardless, this PR adds both of your names to Galaxy's CONTRIBUTORS list.

xref https://github.com/galaxyproject/Galaxy-XSD/issues/2

JeanFred commented:
Wow, I’m really, really happy to see this integrated into core Galaxy!

> @JeanFred, @lindenb as co-copyright holders I was hoping to get your explicit permission to re-license this under the terms of Galaxy's license. It is a permissive (not copy-left) license very much in the spirit of the original MIT license I believe. Regardless, this PR adds both of your names to Galaxy's CONTRIBUTORS list.

Sure, I am fine with re-licensing or dual-licensing my contributions to make it easier to manage for the Galaxy project. I was not very familiar with AFL but it looks good to me. To make it official sounding:

> I hereby agree to license my contributions under the Academic Free License version 3.0 

Thanks everyone!
